components/pd_client/src/client_v2.rs

```diff
@@ -0,0 +1,1418 @@
+            .members
+            .get_header()
+            .get_cluster_id()
+    }
+
+    /// Might panic if `wait_for_ready` isn't called up-front.
+    #[inline]
+    fn header(&self) -> pdpb::RequestHeader {
+        let mut header = pdpb::RequestHeader::default();
+        header.set_cluster_id(self.cluster_id());
+        header
+    }
+
+    #[inline]
+    fn env(&self) -> &Arc<Environment> {
+        &self.core.context.connector.env
+    }
+
+    /// Might panic if `wait_for_ready` isn't called up-front.
+    #[cfg(feature = "testexport")]
+    #[inline]
+    fn leader(&self) -> pdpb::Member {
+        self.cache.as_ref().unwrap().members.get_leader().clone()
+    }
+}
+
+async fn reconnect_loop(
+    mut client: CachedRawClient,
+    cfg: Config,
+    mut should_reconnect: mpsc::Receiver<u64>,
+) {
+    if let Err(e) = client.connect().await {
+        error!("failed to connect pd"; "err" => ?e);
+        return;
+    }
+    loop {
+        if client.channel().wait_for_connected(REQUEST_TIMEOUT).await {
+            let state = ConnectivityState::GRPC_CHANNEL_READY;
+            select! {
+                // Checks for leader change periodically.
+                _ = client
+                    .channel()
+                    .wait_for_state_change(state, cfg.update_interval.0)
+                    .fuse() => {}
+                v = should_reconnect.next().fuse() => {
+                    match v {
+                        None => break,
+                        Some(v) if v < client.cache_version => continue,
+                        _ => {}
+                    }
+                }
+            }
+        }
+        if let Err(e) = client.reconnect().await {
+            warn!("failed to reconnect pd"; "err" => ?e);
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct RpcClient {
+    raw_client: CachedRawClient,
+    feature_gate: FeatureGate,
+}
+
+impl RpcClient {
+    pub fn new(
+        cfg: &Config,
+        shared_env: Option<Arc<Environment>>,
+        security_mgr: Arc<SecurityManager>,
+    ) -> Result<RpcClient> {
+        let env = shared_env.unwrap_or_else(|| {
+            Arc::new(
+                EnvBuilder::new()
+                    .cq_count(CQ_COUNT)
+                    .name_prefix(thd_name!(CLIENT_PREFIX))
+                    .build(),
+            )
+        });
+
+        let (tx, rx) = mpsc::unbounded(mpsc::WakePolicy::Immediately);
+        let raw_client = CachedRawClient::new(cfg.clone(), env, security_mgr, tx);
+        raw_client
+            .new_lame_client()
+            .spawn(reconnect_loop(raw_client.clone(), cfg.clone(), rx));
+
+        Ok(Self {
+            raw_client,
+            feature_gate: Default::default(),
+        })
+    }
+
+    #[inline]
+    pub fn subscribe_reconnect(&self) -> broadcast::Receiver<()> {
+        self.raw_client.clone().on_reconnect_rx
+    }
+
+    #[cfg(feature = "testexport")]
+    pub fn set_on_reconnect<F: Fn() + Sync + Send + 'static>(&self, f: F) {
```

Comment:
Why not change the test to use `subscribe_reconnect`?

Reply:
It's not as convenient because each time the test has to create a runtime/thread itself.

Reply:
> has to create a runtime/thread itself.

Why not just `futures::executor::block_on`?